### Analyzing "Approval" Voting Type from Snapshot Database

1. **Import Necessary Libraries**
```python
import pandas as pd
import json
from sqlalchemy import create_engine
import statsmodels.api as sm
```
- `pandas`: Used for data manipulation and analysis.
- `json`: Used to parse and manipulate JSON formatted data.
- `sqlalchemy`: Allows SQL-based database connections.
- `statsmodels`: Provides classes and functions for statistical models.

2. **Establish Database Connection**
```python
engine = create_engine("mysql+pymysql://root:password@localhost/snapshot_database")
```
- This line creates a connection to the MySQL database. Adjust the connection string if necessary.

3. **Fetch Data from Database**
```python
votes = pd.read_sql('SELECT * FROM votes', con=engine)
proposals = pd.read_sql('SELECT * FROM proposals', con=engine)
```
- Fetches the `votes` and `proposals` tables from the database.

4. **Merge Votes and Proposals Data**
```python
votes_proposals = votes.merge(proposals, left_on='proposal', right_on='id', suffixes=('_vote', '_proposal'))
```
- This merges the `votes` and `proposals` data on the proposal ID.

5. **Filter Data for "Approval" Voting Type**
```python
votes_approval = votes_proposals[votes_proposals['type'] == 'approval'].copy()
```
- Filters the data to only include rows where the voting type is "approval".

6. **Explode Choices for Counting Approvals**
```python
votes_approval_exploded = votes_approval.assign(choice=votes_approval['choice'].str.split(',')).explode('choice')
```
- Splits the comma-separated choices in the `choice` column and creates a new row for each choice.

7. **Determine Winning Choice for Each Proposal**
```python
winning_choices_approval = votes_approval_exploded.groupby('id_proposal')['choice'].value_counts().groupby(level=0).idxmax().apply(lambda x: x[1])
votes_approval['winning_choice'] = votes_approval['id_proposal'].map(winning_choices_approval)
```
- Finds the choice with the maximum count for each proposal and maps it as the winning choice.

8. **Identify Aligned Votes**
```python
votes_approval['aligned'] = votes_approval.apply(lambda row: 1 if row['winning_choice'] in row['choice'] else 0, axis=1)
```
- Marks a vote as aligned (`1`) if the voter's choice includes the winning choice, otherwise marks it as not aligned (`0`).

9. **Lag Variable for Previous Alignment**
```python
votes_approval['previous_aligned'] = votes_approval.groupby(['voter', 'space_vote'])['aligned'].shift()
```
- For each voter and DAO, checks if the previous vote was aligned.

10. **Mark Misaligned Previous Votes**
```python
votes_approval['misaligned_previous'] = (votes_approval['previous_aligned'] == 0).astype(int)
```
- If the previous vote was not aligned, it is marked as misaligned.

11. **Determine Future Voting Behavior**
```python
votes_approval['future_voting'] = votes_approval.groupby(['voter', 'space_vote'])['choice'].shift(-1).notna().astype(int)
```
- Checks if a voter voted in a subsequent proposal within the same DAO.

12. **Regression Analysis**
```python
X = votes_approval[['misaligned_previous']]
X = sm.add_constant(X)  # Adds a constant term to the predictor
y = votes_approval['future_voting']

model = sm.Logit(y, X)
result = model.fit()
print(result.summary())
```
- Performs a logistic regression to analyze the relationship between a misaligned previous vote and the likelihood of voting in a future proposal.

In [1]:
import pandas as pd
import json
from sqlalchemy import create_engine
import statsmodels.api as sm

# Create an engine to the database
engine = create_engine("mysql+pymysql://root:password@localhost/snapshot_database")

# Read in the votes and proposals tables
votes = pd.read_sql('SELECT * FROM votes', con=engine)
proposals = pd.read_sql('SELECT * FROM proposals', con=engine)

# Merge votes and proposals
votes_proposals = votes.merge(proposals, left_on='proposal', right_on='id', suffixes=('_vote', '_proposal'))

# Filter to only "Approval" voting
votes_approval = votes_proposals[votes_proposals['type'] == 'approval'].copy()

# Explode the choices for counting approvals
votes_approval_exploded = votes_approval.assign(choice=votes_approval['choice'].str.split(',')).explode('choice')

# Determine winning choice for each proposal
winning_choices_approval = votes_approval_exploded.groupby('id_proposal')['choice'].value_counts().groupby(level=0).idxmax().apply(lambda x: x[1])
votes_approval['winning_choice'] = votes_approval['id_proposal'].map(winning_choices_approval)

# Determine if a vote was aligned with the winning choice
votes_approval['aligned'] = votes_approval.apply(lambda row: 1 if row['winning_choice'] in row['choice'] else 0, axis=1)

# Create a lag variable for previous alignment
votes_approval['previous_aligned'] = votes_approval.groupby(['voter', 'space_vote'])['aligned'].shift()

# Indicate if the previous vote was misaligned
votes_approval['misaligned_previous'] = (votes_approval['previous_aligned'] == 0).astype(int)

# Indicate if the voter voted in a subsequent proposal within the same DAO
votes_approval['future_voting'] = votes_approval.groupby(['voter', 'space_vote'])['choice'].shift(-1).notna().astype(int)

# Regression analysis
X = votes_approval[['misaligned_previous']]
X = sm.add_constant(X)  # Adds a constant term to the predictor
y = votes_approval['future_voting']

model = sm.Logit(y, X)
result = model.fit()
print(result.summary())

Optimization terminated successfully.
         Current function value: 0.264306
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:          future_voting   No. Observations:               167619
Model:                          Logit   Df Residuals:                   167617
Method:                           MLE   Df Model:                            1
Date:                Wed, 06 Sep 2023   Pseudo R-squ.:                 0.01781
Time:                        07:06:00   Log-Likelihood:                -44303.
converged:                       True   LL-Null:                       -45106.
Covariance Type:            nonrobust   LLR p-value:                     0.000
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                  -2.5817      0.010   -265.946      0.000      -2.601      -2.563
mi

1. **Dep. Variable: `future_voting`**
   - This is the dependent variable you're trying to predict. It indicates whether a voter will vote in a future proposal.

2. **Model: `Logit`**
   - This indicates that a logistic regression model was used.

3. **Pseudo R-squ.: `0.01781`**
   - The Pseudo R-squared value provides a measure of goodness of fit for logistic regression models. In general, a higher value indicates a better fit, but this metric does not have the same interpretation as the R-squared in linear regression. The value you have (about 1.78%) suggests that the model explains a small portion of the variance in the dependent variable. However, in many applications with real-world data, especially in social sciences, even small R-squared values can be meaningful.

4. **Log-Likelihood: `-44303.`**
   - This is the log-likelihood of the model, a measure of fit. The higher the log-likelihood, the better the model fits the data.

5. **LL-Null: `-45106.`**
   - This is the log-likelihood of a model with no predictors, just an intercept. It serves as a comparison to the log-likelihood of the current model.

6. **LLR p-value: `0.000`**
   - This is the p-value for the likelihood ratio test comparing the fit of the current model against the null model. A small p-value (typically < 0.05) indicates that your model fits the data better than a model with no predictors.

7. **Coefficient Table**
   - `const`: This is the intercept. Its coefficient value of `-2.5817` can be interpreted in the log-odds scale. In logistic regression, the intercept is the log-odds of the dependent variable occurring when all predictors are 0.
   - `misaligned_previous`: The coefficient value of `1.4936` indicates the change in the log-odds of `future_voting` for a one-unit change in `misaligned_previous`. Since `misaligned_previous` is binary, this coefficient tells us that the log-odds of voting in the future are about 1.49 times higher for those who had a misaligned vote previously, compared to those who didn't.
   - The `z` values provide a test of the hypothesis that each coefficient is different from 0. A large absolute value of z indicates that you can reject the null hypothesis.
   - The `P>|z|` values test the hypothesis that each coefficient is different from 0. Small p-values (< 0.05) indicate that you can reject the null hypothesis. In this case, both predictors are significant.
   - The `[0.025 0.975]` values are the 95% confidence intervals for the coefficients.

**Key Takeaway**: 
The results suggest that having a misaligned vote in the past is a significant predictor for future voting behavior. Specifically, those who previously voted in a way that was not aligned with the majority are more likely to vote in future proposals compared to those who didn't have a misaligned vote. This provides valuable insights into voting behavior in the context of "Approval" voting type.

In [2]:
num_observations = votes_approval.shape[0]
print(f"Number of observations used in the regression: {num_observations}")


Number of observations used in the regression: 167619
